### PR TITLE
iRODS class for nice printing of ils() output and new docs with working examples

### PR DESCRIPTION
**Pull request by MartinSchobben**
This will be the last one for now. This includes:

- New printing for `ils()`, see here for a preview: https://martinschobben.github.io/irods_client_library_rirods/reference/imeta.html
- This example also shows the second improvement and that is that examples in the manual are now actually working. This is actually an important point of scrutiny during CRAN review. So I hope this helps. Although this will not run on the CRAN server the `is_irods_demo_running()`  predicate ensures that it can be run under the right conditions (a running Docker demo server).
- Lastly, everything is now as good as it gets for a CRAN submission

So I think it is a good time to go for the next step, if we can agree on these changes. It was a still a long haul after starting at the end of last year's summer. 

That said the submission process might still be a long process. So, the more it would make sense to have a go with submission at this point in development, and see what other pitfalls might pop-up.

### CODE REVIEW

File: R/irods-print.R [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+#' # iRODS Zone with metadata
+#' irods_zone <- ils(metadata = TRUE)
+#'
+#' # print (default no row.names)
+#' print(irods_zone)
+#'
+#' # with row.names
+#' print(irods_zone, row.names = TRUE)
+#'
+#' #' # delete object
```
Comment (trel) [inline R/irods-print.R:46]: i see an extra `#'` ?

File: R/irods-s3.R [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+#' irods_zone <- ils(metadata = TRUE)
+#'
+#' # check class
+#' class(irods_zone)
+#'
+#' # coerce into `data.frame` and extract metadata of "foo.rds"
+#' irods_zone <- as.data.frame(irods_zone)
+#' irods_zone[basename(irods_zone$logical_path) == "foo.rds", "metadata"]
+#'
+#' #' # delete object
```
Comment (trel) [inline R/irods-s3.R:51]: same?

File: R/collections.R [hunk 3]
```diff
@@ -56,48 +67,65 @@ irm <- function(logical_path, force = TRUE, recursive = FALSE,
...
+#' unlink(paste0(basename(getwd()), ".irods"))
+#'
+imkdir <- function(logical_path, create_parent_collections = FALSE, verbose = FALSE) {
 
   # expand logical path to absolute logical path
   logical_path <- get_absolute_lpath(logical_path)
 
   # flags to curl call
+  # `imkdir()` will call this API end-point with collection is 1 to create new
+  # collection, 0 does have no affect currently
```
Comment (trel) [inline R/collections.R:124]: affect -> effect

File: R/irods-auth.R [hunk 4]
```diff
@@ -9,9 +19,13 @@ find_irods_file <- function(what) {
 
   # find irods file
   irods <- list.files(".", "\\.irods$")
-  if (length(irods) == 0)
+  if (length(irods) == 0) {
     stop("Can't connect with iRODS server. Did you supply the correct host",
          "name with `create_irods()`?", call. = FALSE)
+  } else if (length(irods) > 1) {
+    stop("You appear to have more than one iRODS project file. ",
+         "Please delete one before proceding.", call. = FALSE)
```
Comment (trel) [inline R/irods-auth.R:27]: proceding -> proceeding

File: R/irods-demo.R [hunk 5]
```diff
@@ -48,41 +46,39 @@ use_irods_demo <- function(user = character(), pass = character(),
...
       !all(grepl(paste0(irods_images_ref, collapse = "|"), irods_images))) {
-    resp_user <- utils::askYesNo(
-      paste0(
-        "The iRODS demo Docker containers are not build on this system.",
-        "Would you like it to be build?"
-      )
-    )
+    message("\nThe iRODS demo Docker containers are not build on this system. \n")
+    resp_user <-
+      utils::askYesNo("Would you like it to be build?", default = FALSE)
```
Comment (trel) [inline R/irods-demo.R:69]: build -> built

File: R/irods-demo.R [hunk 6]
```diff
@@ -48,41 +46,39 @@ use_irods_demo <- function(user = character(), pass = character(),
...
-  if (isTRUE(irods_status) ||
+  if (length(irods_images) == 0 ||
       !all(grepl(paste0(irods_images_ref, collapse = "|"), irods_images))) {
-    resp_user <- utils::askYesNo(
-      paste0(
-        "The iRODS demo Docker containers are not build on this system.",
-        "Would you like it to be build?"
-      )
-    )
+    message("\nThe iRODS demo Docker containers are not build on this system. \n")
```
Comment (trel) [inline R/irods-demo.R:67]: build -> built

File: _pkgdown.yml [hunk 7]
```diff
@@ -2,3 +2,54 @@ url: https://irods.github.io/irods_client_library_rirods/
...
+- title: "Navigating the iRODS Zone"
+  desc: >
+    Logical paths of iRODS and listing content of collections
+  contents:
+  - icd
+  - ipwd
+  - ils
+- title: "Managing collections"
+  desc: >
+    Creating and removing data objects and collections withtin iRODS
```
Comment (trel) [inline _pkgdown.yml:23]: withtin -> within

File: R/irods-print.R (hunk 1 above)
Comment (MartinSchobben) [reply R/irods-print.R]: done

File: R/irods-s3.R (hunk 2 above)
Comment (MartinSchobben) [reply R/irods-s3.R]: done

File: R/collections.R (hunk 3 above)
Comment (MartinSchobben) [reply R/collections.R]: done

File: R/irods-auth.R (hunk 4 above)
Comment (MartinSchobben) [reply R/irods-auth.R]: done

File: R/irods-demo.R (hunk 5 above)
Comment (MartinSchobben) [reply R/irods-demo.R]: done

File: R/irods-demo.R (hunk 6 above)
Comment (MartinSchobben) [reply R/irods-demo.R]: done

File: _pkgdown.yml (hunk 7 above)
Comment (MartinSchobben) [reply _pkgdown.yml]: done

File: R/admin.R [hunk 8]
```diff
@@ -17,18 +17,31 @@
 #' @return Invisible http status.
 #' @export
 #'
-#' @examples
-#' if (interactive()) {
-#'   # add user
-#'   iadmin(action = "add", target = "user", arg2 = "Bob", arg3 = "rodsuser")
+#' @examplesIf is_irods_demo_running()
```
Comment (montesmariana) [inline R/admin.R:20]: ? Space or new line missing?

File: R/authentication.R [hunk 9]
```diff
@@ -11,15 +11,26 @@
 #' @return Invisibly `NULL`.
 #' @export
 #'
-#' @examples
-#' if(interactive()) {
+#' @examplesIf is_irods_demo_running()
```
Comment (montesmariana) [inline R/authentication.R:14]: Same as above?

File: R/data-objects.R [hunk 10]
```diff
@@ -205,15 +221,16 @@ local_to_irods_ <- function(
   irods_rest_call("stream", "PUT", args, verbose, object)
 }
 
-#' Retrieve file or object from iRODS
+#' Retrieve File or Object from iRODS
 #'
 #' Transfer a file from iRODS to the local storage with `iget()` or
-#' read an R object from an RDS file in iRODS with `ireadRDS()` (see `readRDS()`).
+#' read an R object from an RDS file in iRODS with `ireadRDS()`
```
Comment (montesmariana) [inline R/data-objects.R:227]: `[ireadRDS()]` instead of `` `ireadRDS()` ``

File: R/data-objects.R [hunk 11]
```diff
@@ -19,28 +19,44 @@
 #'  exists. Defaults to `FALSE`.
 #'
 #' @return (Invisibly) the HTTP response.
+#' @seealso
+#'  [iget()] for obtaining files,
+#'  [ireadRDS] for obtaining R objects from iRODS,
```
Comment (montesmariana) [inline R/data-objects.R:24]: Add parentheses: `[ireadRDS()]` and `[saveRDS()]`

File: R/data-objects.R [hunk 12]
```diff
@@ -19,28 +19,44 @@
...
+#'  [saveRDS] for an R equivalent.
 #' @export
 #'
-#' @examples
-#' if (interactive()) {
+#' @examplesIf is_irods_demo_running()
```
Comment (montesmariana) [inline R/data-objects.R:28]: Space?

File: R/data-objects.R [hunk 13]
```diff
@@ -223,31 +240,50 @@ local_to_irods_ <- function(
 #'
 #' @return The R object in case of `ireadRDS()`, invisibly `NULL` in case of
 #'  `iget()`.
+#' @seealso
+#'  [iput()] for sending files,
+#'  [isaveRDS] for sending R objects to iRODS,
```
Comment (montesmariana) [inline R/data-objects.R:245]: here and next line: add parentheses (`[isaveRDS()]` and `[readRDS()]`)

File: R/data-objects.R [hunk 14]
```diff
@@ -223,31 +240,50 @@ local_to_irods_ <- function(
...
+#'  [readRDS] for an R equivalent.
+#'
 #' @export
 #'
-#' @examples
-#' if (interactive()) {
-#' create_irods("http://localhost/irods-rest/0.9.3", "/tempZone/home", TRUE)
+#' @examplesIf is_irods_demo_running()
```
Comment (montesmariana) [inline R/data-objects.R:250]: space?

File: R/irods-demo.R [hunk 15]
```diff
@@ -1,9 +1,7 @@
-#' Run Docker iRODS demonstration service
+#' Run Docker iRODS Demonstration Service
 #'
-#' Run an iRODS demonstration server as an Docker container instance.
-#' The functions `stop_irods_demo()` and `remove_docker_images()` are used to
-#' stop the containers, and subsequently remove the images. Note that calling
-#' `use_irods_demo()` consecutively will reset the server with a clean slate.
+#' Run an iRODS demonstration server with `use_irods_demo()` as an Docker
```
Comment (montesmariana) [inline R/irods-demo.R:3]: a Docker container...

File: R/irods-rest.R [hunk 16]
```diff
@@ -1,3 +1,15 @@
+#' Internal function to coordinate iRODS REST call
+#'
+#' @param endpoint REST API end-point.
+#' @param verb The HTTP verb to be used (e.g. PUT).
+#' @param args The arguments to be set for the end-point.
+#' @param verbose Verbosity of the HTTP request.
+#' @param object Data to send along with HTTP request.
+#' @param error Should status codes >400 be translated to errors.
```
Comment (montesmariana) [inline R/irods-rest.R:8]: (In alignment with other boolean args around the package)
"Whether status codes larger than 400 should be translated to errors."

File: R/irods-s3.R [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+#' @param x `irods_df` class object.
+#' @param ... Currently not implemented
+#'
+#' @return Returns a `data.frame`. Note, that the
+#' columns of metadata consists of a list of data frames, and status_information
+#' and permission_information consist of data frames.
+#'
+#' @export
+#'
+#' @examplesIf is_irods_demo_running()
```
Comment (montesmariana) [inline R/irods-s3.R:14]: Space

File: R/metadata.R [hunk 18]
```diff
@@ -9,30 +9,35 @@
...
+#' @seealso [iquery()]
 #'
 #' @references
 #' https://docs.irods.org/master/icommands/metadata/
 #'
 #' @export
 #'
-#' @examples
-#' if(interactive()) {
+#' @examplesIf is_irods_demo_running()
```
Comment (montesmariana) [inline R/metadata.R:30]: Space

**Review by montesmariana — APPROVED** [review]
Looks great!
I remarked a few things in docs, but minimal. I didn't know the `examplesIf` tag (interesting!) so I commented on it and then I removed the comments but there might be leftovers if I didn't clean it properly.

File: R/authentication.R [hunk 19]
```diff
@@ -1,4 +1,4 @@
-#' Authentication service for the iRODS zone
+#' Authentication Service for the iRODS Zone
```
[the remark this answers is not in the document]
Comment (MartinSchobben) [reply R/authentication.R]: changed to "an"

File: R/data-objects.R (hunk 10 above)
Comment (MartinSchobben) [reply R/data-objects.R]: done

File: R/data-objects.R (hunk 11 above)
Comment (MartinSchobben) [reply R/data-objects.R]: done

File: R/data-objects.R (hunk 13 above)
Comment (MartinSchobben) [reply R/data-objects.R]: done

File: R/irods-demo.R (hunk 15 above)
Comment (MartinSchobben) [reply R/irods-demo.R]: changed

File: R/irods-rest.R (hunk 16 above)
Comment (MartinSchobben) [reply R/irods-rest.R]: changed as well

**Comment by MartinSchobben** [comment]
@trel, I imagine that you are busy but if you still find the time then please merge. Then my talk will be in line with main.

**Pull request merged**